sklearn/utils/_namedarray.py

```diff
@@ -0,0 +1,103 @@
+# -*- coding: utf-8 -*-
+# Authors: Adrin Jalali <adrin.jalali@gmail.com>
+#
+# License: BSD 3 clause
+
+import numpy as np
+
+# NDArrayOperatorsMixin was added in numpy 1.13
+# TODO: cleanup once we support numpy 1.13+
+try:
+    from numpy.lib.mixins import NDArrayOperatorsMixin
+except ImportError:
+    raise NotImplementedError("In order to use NamedAraay, please upgrade your "
+                              "numpy to 1.13+!")
+
+from .validation import check_array, column_or_1d
+
+
+class NamedArray(NDArrayOperatorsMixin):
+    """A wrapper to a numpy ndarray holding some metadata about the data.
+
+    Instances of this object behave like a numpy array, and loose all metadata
+    information in numerical operations.
+
+    Parameters
+    ----------
+    data: array-like
+        A one or two dimensional array like data.
+
+    feature_names: list or array of strings, or None, default=None
+        Feature names associated with the columns of the data. The number of
+        columns should always be the same as the number of feature names.
+        Setting the `data` of an instance, would result in `feature_names` to
+        be `None` if the number of columns do not match the number of stored
+        feature names.
+    """
+    _feature_names = None
+    _data = None
+
+    def __init__(self, data, feature_names=None):
+        self.data = data
+        self.feature_names = feature_names
+
+    @property
+    def data(self):
```

Comment:
NumPy arrays do actually have a rarely used `.data` attribute (which returns a buffer object). I'm not sure that you want to override that publicly -- it might be enough to treat `_data` as immutable (and private) and say "Use `np.asarray()` to convert without a copy."